### PR TITLE
WIP: Cookie Impl

### PR DESCRIPTION
**Pull request by Shabinder**
PR building Cookie Management in client using WebView.

### CODE REVIEW

**Comment by Shabinder** [comment]
@KevinnZou 
I have setup a blueprint for the cookie manager feature, you have been invited to the fork and given access to push changes from your side as well.

I will be having somewhat super packed up coming weeks, so you are more than welcome to take this to end stage in case I might take extended time. Appreciate your efforts in here.

**Comment by KevinnZou** [comment]
> @KevinnZou I have setup a blueprint for the cookie manager feature, you have been invited to the fork and given access to push changes from your side as well.
> 
> I will be having somewhat super packed up coming weeks, so you are more than welcome to take this to end stage in case I might take extended time. Appreciate your efforts in here.

Thank you for providing the blueprint setup. It will be very helpful to me. Last week, we focused on transitioning from JavaFX to the CEF browser. We were able to solve the issue today and we plan to release the new version tomorrow. After that, we will have time to address this issue and aim to implement it by the end of next week. We will keep you updated on our progress. Thank you once again for your contribution!

File: webview/src/commonMain/kotlin/com/multiplatform/webview/cookie/Cookie.kt [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+
+        if (sameSite != null) cookieValue += "; SameSite=$sameSite"
+
+        cookieValue += ";"
+
+        return cookieValue
+    }
+}
+
+expect fun getCookieExpirationDate(expiresDate: Long): String
```
Comment (KevinnZou) [inline webview/src/commonMain/kotlin/com/multiplatform/webview/cookie/Cookie.kt:101]: would it be better to just use the [kotlinx-datetime library](https://github.com/Kotlin/kotlinx-datetime)?

Comment (Shabinder) [reply webview/src/commonMain/kotlin/com/multiplatform/webview/cookie/Cookie.kt:101]: We can if it increases convenience, if its just this usage and no usage in future then adding additional library doesnt make much sense.

Comment (KevinnZou) [reply webview/src/commonMain/kotlin/com/multiplatform/webview/cookie/Cookie.kt:101]: makes sense.

**Comment by KevinnZou** [comment]
@Shabinder I have implemented CookieManager for iOS and Desktop. It works well on iOS, but there is a major issue on Desktop.

The problem is that I am using the CefCookieManager obtained through `CefCookieManager.getGlobalManager()` to manage cookies. However, it doesn't seem to work. Whether it's `setCookie` or `visitUrlCookies`, the method calls have no effect.

After some research, I found an existing [issue](https://github.com/chromiumembedded/java-cef/issues/405) in the JCEF repository that is related to this problem. @DATL4G, could you please take a look at it? If the issue with the CefCookieManager cannot be resolved, we may need to explore alternative methods to support cookie management on the Desktop platform.

File: webview/src/commonMain/kotlin/com/multiplatform/webview/cookie/CookieManager.kt [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+ * Cookie Manager exposing access to cookies of the WebView.
+ * This is an interface to allow platform specific implementations.
+ * ---------------------------------------------------------------
+ * PS: Not having it as expect/actual class was a conscious decision,
+ * since expect/actual classes will be marked as beta in coming kotlin releases.
+ * */
+interface CookieManager {
+    suspend fun setCookie(url: String, cookie: Cookie)
+    suspend fun getCookies(url: String): List<Cookie>
+    suspend fun removeAllCookies()
```
Comment (Shabinder) [inline webview/src/commonMain/kotlin/com/multiplatform/webview/cookie/CookieManager.kt:14]: We should also have a delete cookie function for a URL.

`suspend fun removeCookies(url: String)`

Comment (Shabinder) [reply webview/src/commonMain/kotlin/com/multiplatform/webview/cookie/CookieManager.kt:14]: Also `WebSettings` should have following options.
```

  - isCookiesEnabled: Boolean
  - isThirdPartyCookiesEnabled: Boolean
```

Comment (Shabinder) [reply webview/src/commonMain/kotlin/com/multiplatform/webview/cookie/CookieManager.kt:14]: Great progress overall, kudos @KevinnZou

**Comment by DATL4G** [comment]
@Shabinder Could not push to your PR directly, opened a PR https://github.com/Shabinder/compose-webview-multiplatform/pull/1

@KevinnZou 
The GlobalCookieManager works but is handled on the IO thread and may take a while, since it opens a connection to the local cookie database and so on
I added a in-memory save of cookies, may need some improvements but worked as expected on my end.
It will load all cookies added by websites and manually ones

**Comment by Shabinder** [comment]
> @Shabinder Could not push to your PR directly, opened a PR [Shabinder#1](https://github.com/Shabinder/compose-webview-multiplatform/pull/1)

Have merged it, and sent u an invite for access aswell.

**Comment by KevinnZou** [comment]
@DATL4G I tested your PR locally and it turns out that it still does not work. We can only get the cookie we set before. It is because we have a local cache. But for other cookies sent by the website, we still cannot get it.

**Comment by DATL4G** [comment]
@KevinnZou I think we need to stall this on desktop then.

I can write a kotlin library that handles the native bindings etc, so basically a newer, better and maybe depending on Jetbrains CEF version of JCEF.
But this takes some weeks

**Comment by DATL4G** [comment]
@KevinnZou we could also read the cookie sqlite file ourself (already tried that with no success until I noticed that we delete all cookies right after adding one 🤦🏻)

**Comment by KevinnZou** [comment]
@DATL4G Thanks for your investigation! If the current version of JCEF does not support it, I think we can release a version that supports Android and iOS first. As for the desktop, I am looking forward to seeing your library that is based on Jetbrains CEF, which will definitely have better performance.

**Comment by DATL4G** [comment]
Please test https://github.com/DATL4G/KCEF by doing the following:

```bash
git clone git@github.com:DATL4G/KCEF.git --recurse-submodules

./gradlew example:run # Or run it from IntelliJ
```

@Shabinder do you use Windows?
I'm using Linux and @KevinnZou uses MacOS, would be nice if someone could test the Windows target as well.

**Comment by Shabinder** [comment]
> Please test https://github.com/DATL4G/KCEF by doing the following:
> 
> ```bash
> git clone git@github.com:DATL4G/KCEF.git --recurse-submodules
> 
> ./gradlew example:run # Or run it from IntelliJ
> ```
> 
> @Shabinder do you use Windows?
> I'm using Linux and @KevinnZou uses MacOS, would be nice if someone could test the Windows target as well.

I am a mac user aswell.

**Comment by DATL4G** [comment]
Tested https://github.com/DATL4G/KCEF on Windows and added some patches.

Added fixes for MacOS as well, @Shabinder @KevinnZou if you tried it before, please do it again.
If you haven't tested it yet, thats fine, as it probably didn't work before.

**Comment by KevinnZou** [comment]
Apologies for the delayed response. I have recently returned from a lengthy vacation. 

Regarding your project, I encountered an issue while testing it on my Mac system. Specifically, it crashed right after the Jcef Bundled was downloaded. Here is the corresponding stack trace:
```
"/Applications/Android Studio.app/Contents/jbr/Contents/Home/bin/java" -javaagent:/Applications/Android Studio.app/Contents/lib/idea_rt.jar=52702:/Applications/Android Studio.app/Contents/bin -Dfile.encoding=UTF-8 -classpath /Users/kevinnzou/NetEase/Compose/fork/datl4g/KCEF/example/build/classes/kotlin/main:/Users/kevinnzou/NetEase/Compose/fork/datl4g/KCEF/kcef/build/classes/kotlin/main:/Users/kevinnzou/.gradle/caches/modules-2/files-2.1/org.jetbrains.kotlin/kotlin-stdlib-jdk8/1.9.0/e000bd084353d84c9e888f6fb341dc1f5b79d948/kotlin-stdlib-jdk8-1.9.0.jar:/Users/kevinnzou/NetEase/Compose/fork/datl4g/KCEF/jcef/build/classes/java/main:/Users/kevinnzou/.gradle/caches/modules-2/files-2.1/org.jetbrains.kotlin/kotlin-stdlib-jdk7/1.9.0/f320478990d05e0cfaadd74f9619fd6027adbf37/kotlin-stdlib-jdk7-1.9.0.jar:/Users/kevinnzou/.gradle/caches/modules-2/files-2.1/org.jetbrains.kotlin/kotlin-stdlib/1.9.0/8ee15ef0c67dc83d874f412d84378d7f0eb50b63/kotlin-stdlib-1.9.0.jar:/Users/kevinnzou/.gradle/caches/modules-2/files-2.1/org.jetbrains.compose.desktop/desktop-jvm/1.5.2/3b461ec5c3f11ec2a6b169f6c7290415f7b60bf3/desktop-jvm-1.5.2.jar:/Users/kevinnzou/.gradle/caches/modules-2/files-2.1/org.jetbrains.kotlin/kotlin-stdlib-common/1.9.0/cd65c21cfd1eec4d44ef09f9f52b6d9f8a720636/kotlin-stdlib-common-1.9.0.jar:/Users/kevinnzou/.gradle/caches/modules-2/files-2.1/org.jetbrains.compose.material/material-desktop/1.5.2/99ad414ebbfd5999dbc093853362c81288588701/material-desktop-1.5.2.jar:/Users/kevinnzou/.gradle/caches/modules-2/files-2.1/org.jetbrains.compose.foundation/foundation-desktop/1.5.2/8e6625123de5362540bd3a6e6b438b06118be85c/foundation-desktop-1.5.2.jar:/Users/kevinnzou/.gradle/caches/modules-2/files-2.1/org.jetbrains.compose.ui/ui-desktop/1.5.2/adc0dfcf7a8731a0a680b9e4d1f3e3d3d9917493/ui-desktop-1.5.2.jar:/Users/kevinnzou/.gradle/caches/modules-2/files-2.1/org.jetbrains.compose.ui/ui-tooling-preview-desktop/1.5.2/16e8e3a0dd0d52b5e4877a54bc345668ba23428d/ui-tooling-preview-desktop-1.5.2.jar:/Users/kevinnzou/.gradle/caches/modules-2/files-2.1/org.jetbrains.compose.runtime/runtime-desktop/1.5.2/750cae784e9db6ae3e00c52489269febe42f087c/runtime-desktop-1.5.2.jar:/Users/kevinnzou/.gradle/caches/modules-2/files-2.1/org.jetbrains.compose.material/material-ripple-desktop/1.5.2/494966d7ed5d7143b7e67f10077f8f20ab9fa1ec/material-ripple-desktop-1.5.2.jar:/Users/kevinnzou/.gradle/caches/modules-2/files-2.1/org.jetbrains.compose.material/material-icons-core-desktop/1.5.2/82faf5401b15358511cb941fd1306433ae12be77/material-icons-core-desktop-1.5.2.jar:/Users/kevinnzou/.gradle/caches/modules-2/files-2.1/org.jetbrains.compose.ui/ui-text-desktop/1.5.2/44519ce6b5b12b70308746ed4351664fdf428322/ui-text-desktop-1.5.2.jar:/Users/kevinnzou/.gradle/caches/modules-2/files-2.1/org.jetbrains.compose.animation/animation-core-desktop/1.5.2/c9cbe3824dd25e0152e9298574d24ec9ac61d016/animation-core-desktop-1.5.2.jar:/Users/kevinnzou/.gradle/caches/modules-2/files-2.1/org.jetbrains.compose.animation/animation-desktop/1.5.2/f1cbe23da9ecb37aa6b2b3e57b223327227b910a/animation-desktop-1.5.2.jar:/Users/kevinnzou/.gradle/caches/modules-2/files-2.1/org.jetbrains.skiko/skiko-awt/0.7.77/e6e5eb033a8d1507d71e0817f64e1e93fc7ec9ef/skiko-awt-0.7.77.jar:/Users/kevinnzou/.gradle/caches/modules-2/files-2.1/org.jetbrains.compose.ui/ui-graphics-desktop/1.5.2/694e3a460d5fc15245b2502ae4025113f132abb2/ui-graphics-desktop-1.5.2.jar:/Users/kevinnzou/.gradle/caches/modules-2/files-2.1/org.jetbrains.compose.runtime/runtime-saveable-desktop/1.5.2/7afa940f340c1432bf95466463cdc00baa6e63eb/runtime-saveable-desktop-1.5.2.jar:/Users/kevinnzou/.gradle/caches/modules-2/files-2.1/org.jetbrains.compose.ui/ui-unit-desktop/1.5.2/ce79e803d4f0b33631c958f64ff36a7010580d16/ui-unit-desktop-1.5.2.jar:/Users/kevinnzou/.gradle/caches/modules-2/files-2.1/org.jetbrains.compose.ui/ui-geometry-desktop/1.5.2/81ea1ab414ba9926972708e9f4ca7ac499cd9137/ui-geometry-desktop-1.5.2.jar:/Users/kevinnzou/.gradle/caches/modules-2/files-2.1/org.jetbrains.compose.foundation/foundation-layout-desktop/1.5.2/ae3ea49f6a2e5a0a6d4996c2bc10c3ddeb1f5dd5/foundation-layout-desktop-1.5.2.jar:/Users/kevinnzou/.gradle/caches/modules-2/files-2.1/org.apache.commons/commons-compress/1.24.0/b4b1b5a3d9573b2970fddab236102c0a4d27d35e/commons-compress-1.24.0.jar:/Users/kevinnzou/.gradle/caches/modules-2/files-2.1/org.jetbrains.skiko/skiko-awt-runtime-macos-x64/0.7.77/758fe43505f09e982c080fe52df2bbf978e685b0/skiko-awt-runtime-macos-x64-0.7.77.jar:/Users/kevinnzou/.gradle/caches/modules-2/files-2.1/org.jogamp.gluegen/gluegen-rt/2.5.0/683e9d628c2ca32c4b56249034e164e593bf37b3/gluegen-rt-2.5.0.jar:/Users/kevinnzou/.gradle/caches/modules-2/files-2.1/org.jogamp.jogl/jogl-all/2.5.0/68dc55e290bd5456e87f5d47834de47633dbe754/jogl-all-2.5.0.jar:/Users/kevinnzou/.gradle/caches/modules-2/files-2.1/io.ktor/ktor-client-okhttp-jvm/2.3.4/93a5f3aa1a2ef750e5a3a95a27576d2e5c6bea21/ktor-client-okhttp-jvm-2.3.4.jar:/Users/kevinnzou/.gradle/caches/modules-2/files-2.1/io.ktor/ktor-client-content-negotiation-jvm/2.3.4/acb20f3b68ae0c265f16607ffd03a3ec89eb0f99/ktor-client-content-negotiation-jvm-2.3.4.jar:/Users/kevinnzou/.gradle/caches/modules-2/files-2.1/io.ktor/ktor-client-core-jvm/2.3.4/7e6b050441579959bf26d2f07143099809d7b276/ktor-client-core-jvm-2.3.4.jar:/Users/kevinnzou/.gradle/caches/modules-2/files-2.1/io.ktor/ktor-serialization-kotlinx-json-jvm/2.3.4/4d34f83abd1202a0daa7cfa464a25f2db3606b08/ktor-serialization-kotlinx-json-jvm-2.3.4.jar:/Users/kevinnzou/.gradle/caches/modules-2/files-2.1/org.jetbrains.kotlinx/kotlinx-coroutines-core-jvm/1.7.3/2b09627576f0989a436a00a4a54b55fa5026fb86/kotlinx-coroutines-core-jvm-1.7.3.jar:/Users/kevinnzou/.gradle/caches/modules-2/files-2.1/org.jetbrains/annotations/23.0.0/8cc20c07506ec18e0834947b84a864bfc094484e/annotations-23.0.0.jar:/Users/kevinnzou/.gradle/caches/modules-2/files-2.1/antlr/antlr/2.7.7/83cd2cd674a217ade95a4bb83a8a14f351f48bd0/antlr-2.7.7.jar:/Users/kevinnzou/.gradle/caches/modules-2/files-2.1/org.jetbrains.kotlinx/kotlinx-coroutines-jdk8/1.7.3/263611b87f7546402852e43a8dc7ef3223ada42a/kotlinx-coroutines-jdk8-1.7.3.jar:/Users/kevinnzou/.gradle/caches/modules-2/files-2.1/com.squareup.okhttp3/okhttp/4.11.0/436932d695b2c43f2c86b8111c596179cd133d56/okhttp-4.11.0.jar:/Users/kevinnzou/.gradle/caches/modules-2/files-2.1/org.slf4j/slf4j-api/1.7.36/6c62681a2f655b49963a5983b8b0950a6120ae14/slf4j-api-1.7.36.jar:/Users/kevinnzou/.gradle/caches/modules-2/files-2.1/org.jetbrains.kotlinx/kotlinx-coroutines-slf4j/1.7.3/34086fbfd556969d86b45ee2a69b816ee2413701/kotlinx-coroutines-slf4j-1.7.3.jar:/Users/kevinnzou/.gradle/caches/modules-2/files-2.1/com.squareup.okio/okio-jvm/3.4.0/4e8bd78a52ab935ce383d0092646922154295e54/okio-jvm-3.4.0.jar:/Users/kevinnzou/.gradle/caches/modules-2/files-2.1/io.ktor/ktor-serialization-jvm/2.3.4/b246a515c777d23e12abb7824773333cc41357f0/ktor-serialization-jvm-2.3.4.jar:/Users/kevinnzou/.gradle/caches/modules-2/files-2.1/io.ktor/ktor-events-jvm/2.3.4/1d08d8375a808a2832f3a5f52c8e088c1ab2c86f/ktor-events-jvm-2.3.4.jar:/Users/kevinnzou/.gradle/caches/modules-2/files-2.1/io.ktor/ktor-websocket-serialization-jvm/2.3.4/a6d6a02d434ca6894655f1f3c199c1cd7b8a5b2e/ktor-websocket-serialization-jvm-2.3.4.jar:/Users/kevinnzou/.gradle/caches/modules-2/files-2.1/io.ktor/ktor-http-jvm/2.3.4/dc1b1cb0e6ef918c8a432106a81a10b72304b58e/ktor-http-jvm-2.3.4.jar:/Users/kevinnzou/.gradle/caches/modules-2/files-2.1/io.ktor/ktor-serialization-kotlinx-jvm/2.3.4/f7c20edfc5e3af71c0d3679be1f36d76afed4096/ktor-serialization-kotlinx-jvm-2.3.4.jar:/Users/kevinnzou/.gradle/caches/modules-2/files-2.1/org.jetbrains.kotlinx/kotlinx-serialization-json-jvm/1.5.1/21ac91884e0b9462b9106d1a83e5e0d200170c65/kotlinx-serialization-json-jvm-1.5.1.jar:/Users/kevinnzou/.gradle/caches/modules-2/files-2.1/org.jetbrains.compose.ui/ui-util-desktop/1.5.2/b6cef817ffe7a8b7bba1ef34ded701f287cebf9/ui-util-desktop-1.5.2.jar:/Users/kevinnzou/.gradle/caches/modules-2/files-2.1/io.ktor/ktor-websockets-jvm/2.3.4/593843969c814e5de2f5cdb64441d6bcd0eeb010/ktor-websockets-jvm-2.3.4.jar:/Users/kevinnzou/.gradle/caches/modules-2/files-2.1/io.ktor/ktor-utils-jvm/2.3.4/ac6c349dbbe6c6f7dbe6f522531c2fad03733ef6/ktor-utils-jvm-2.3.4.jar:/Users/kevinnzou/.gradle/caches/modules-2/files-2.1/org.jetbrains.kotlinx/kotlinx-serialization-core-jvm/1.5.1/e26cf5dfbcfe3e82ca196694dfd305753b1a49b9/kotlinx-serialization-core-jvm-1.5.1.jar:/Users/kevinnzou/.gradle/caches/modules-2/files-2.1/org.jetbrains.kotlinx/atomicfu-jvm/0.17.0/d3d2e380b6ee28231118101d09ae6dd06a1db1cb/atomicfu-jvm-0.17.0.jar:/Users/kevinnzou/.gradle/caches/modules-2/files-2.1/io.ktor/ktor-io-jvm/2.3.4/9b5c5d64c49b4cd1843bd589cdc0b360ddf11480/ktor-io-jvm-2.3.4.jar MainKt
JCEF(11:48:212): initialized stderr logger, severity=LOGSEVERITY_DEFAULT
#
# A fatal error has been detected by the Java Runtime Environment:
#
#  SIGSEGV (0xb) at pc=0x0000000000000000, pid=14303, tid=65587
#
# JRE version: OpenJDK Runtime Environment (17.0.6) (build 17.0.6+0-17.0.6b829.9-10027231)
# Java VM: OpenJDK 64-Bit Server VM (17.0.6+0-17.0.6b829.9-10027231, mixed mode, tiered, compressed oops, compressed class ptrs, g1 gc, bsd-amd64)
# Problematic frame:
# C  0x0000000000000000
#
# No core dump will be written. Core dumps have been disabled. To enable core dumping, try "ulimit -c unlimited" before starting Java again
#
JNI global refs:
JNI global refs: 145, weak refs: 6

JNI global refs memory usage: 2099, weak refs: 833

# An error report file with more information is saved as:
# /Users/kevinnzou/NetEase/Compose/fork/datl4g/KCEF/hs_err_pid14303.log
#
# If you would like to submit a bug report, please visit:
#   https://bugreport.java.com/bugreport/crash.jsp
# The crash happened outside the Java Virtual Machine in native code.
# See problematic frame for where to report the bug.
#
dlopen /Applications/Android Studio.app/Contents/jbr/Contents/Frameworks/Chromium Embedded Framework.framework/Chromium Embedded Framework: dlopen(/Applications/Android Studio.app/Contents/jbr/Contents/Frameworks/Chromium Embedded Framework.framework/Chromium Embedded Framework, 0x0105): tried: '/Applications/Android Studio.app/Contents/jbr/Contents/Frameworks/Chromium Embedded Framework.framework/Chromium Embedded Framework' (no such file), '/System/Volumes/Preboot/Cryptexes/OS/Applications/Android Studio.app/Contents/jbr/Contents/Frameworks/Chromium Embedded Framework.framework/Chromium Embedded Framework' (no such file), '/Applications/Android Studio.app/Contents/jbr/Contents/Frameworks/Chromium Embedded Framework.framework/Chromium Embedded Framework' (no such file), '/System/Library/Frameworks/Chromium Embedded Framework.framework/Chromium Embedded Framework' (no such file, not in dyld cache)
JCEF_I(11:48:215): CefApp: set state NEW
JCEF_V(11:48:218): Fixed args: [--framework-dir-path=/Applications/Android Studio.app/Contents/jbr/Contents/Frameworks/Chromium Embedded Framework.framework, --browser-subprocess-path=/Applications/Android Studio.app/Contents/jbr/Contents/Frameworks/jcef Helper.app/Contents/MacOS/jcef Helper, --main-bundle-path=/Applications/Android Studio.app/Contents/jbr/Contents/Frameworks/jcef Helper.app]
JCEF_I(11:48:221): CefApp: set state INITIALIZING
JCEF_V(11:48:222): Initialize CefApp on Thread[CefInitialize-thread,6,main]

Process finished with exit code 134 (interrupted by signal 6: SIGABRT)
```

**Comment by DATL4G** [comment]
@KevinnZou thanks for reporting, I applied some fixes.
Please delete the jcef-bundle folder and pull the latest changes

**Comment by KevinnZou** [comment]
> @DATL4G Thanks for your investigation! If the current version of JCEF does not support it, I think we can release a version that supports Android and iOS first. As for the desktop, I am looking forward to seeing your library that is based on JetBrains CEF, which will definitely have better performance.

@DATL4G As previously discussed, I will clean up this MR today and release it with the latest version. Regarding KCEF, Before switching to it, I believe it is necessary to conduct a comprehensive test and wait for public feedback. It may take some time to become stable and ready for use. Let's use this separate [issue](https://github.com/KevinnZou/compose-webview-multiplatform/issues/19) to discuss future problems with KCEF.

**Review by KevinnZou — APPROVED** [review]
Due to some issues with JCEF, we have decided to initially support Android and iOS platforms. Once the KCEF testing is stable, we will switch to it and also support cookies on the Desktop platform.

**Pull request merged**